File: Src/ButterflyNet.Parser/Strategies/StrongStrategy.cs
using System;
using ButterflyNet.Parser.Satisfiers;
using ButterflyNet.Parser.Scopes;

namespace ButterflyNet.Parser.Strategies {
	[TokenTransformer("__")]
	public abstract class StrongStrategy : InlineStrategy {
		protected Type Type { get { return ScopeTypeCache.Strong; } }
	}

	public class OpenStrongStrategy : StrongStrategy {
		public OpenStrongStrategy() {
			AddSatisfier(new OpenNonNestableInlineScopeSatisfier(Type));
		}

		protected override void DoExecute(ParseContext context) {
			OpenScope(new StrongScope(), context);
		}
	}

	public class CloseStrongStrategy : StrongStrategy {
		public CloseStrongStrategy() {
			AddSatisfier(new CurrentScopeMustMatchSatisfier(Type));
		}

		protected override void DoExecute(ParseContext context) {
			CloseCurrentScope(context);
		}
	}
}

File: src/js/src/strategies/StrongStrategy.js
function StrongStrategy() {
	StrongStrategy.$parent.call(this);
	this.type = ScopeTypeCache.strong;
	this.setAsTokenTransformer("__");
}

extend(InlineStrategy, StrongStrategy);

function OpenStrongStrategy() {
	OpenStrongStrategy.$parent.call(this);
	
	this.addSatisfier(new OpenNonNestableInlineScopeSatisfier(this.type));
	
	this.doExecute = function(context) {
		this.openScope(new StrongScope(), context);
	};
}

extend(StrongStrategy, OpenStrongStrategy);

function CloseStrongStrategy() {
	CloseStrongStrategy.$parent.call(this);
	
	this.addSatisfier(new CurrentScopeMustMatchSatisfier(this.type));
	
	this.doExecute = function(context) {
		this.closeCurrentScope(context);
	};
}

extend(StrongStrategy, CloseStrongStrategy);